### PR TITLE
Adds possibility to edit an event in web2

### PR DESCRIPTION
**Pull request by heiglandreas**
This commit adds the possibility to edit an event via the web2 interface

### CODE REVIEW

**Comment by lornajane** [comment]
Right, I've added a PR for the API which gives a can_edit field on an event telling you whether the current user has the right to edit it or not.  Can you let me know what you think? https://github.com/joindin/joindin-api/pull/137

**Comment by heiglandreas** [comment]
@lornajane That's looking great! I'll change the code accordingly

**Comment by akrabat** [comment]
I will look at this once we've sorted out the DateTime issues with the cfp fields on `api`.

**Comment by akrabat** [comment]
I believe this one is waiting on https://github.com/joindin/joindin-api/pull/134

**Comment by heiglandreas** [comment]
It is. :grinning:

File: app/src/Event/EventController.php [hunk 1]
```diff
@@ -307,6 +397,33 @@ private function addEventUsingForm(Form $form)
 
         $result = false;
         try {
+            $result = $eventApi->edit($values);
```
Comment (akrabat) [inline app/src/Event/EventController.php]: This should be $eventApi->submit()

File: app/src/Event/EventController.php [hunk 2]
```diff
@@ -307,6 +397,33 @@ private function addEventUsingForm(Form $form)
...
+     * @return EventEntity|null|false
+     */
+    private function editEventUsingForm(Form $form)
+    {
+        $eventApi = $this->getEventApi();
+        $values = $form->getData()->toArray();
+
+        $result = false;
+        try {
             $result = $eventApi->submit($values);
```
Comment (akrabat) [inline app/src/Event/EventController.php]: This should be $eventApi->edit()

File: app/templates/Event/edit.html.twig [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+{% use '/Event/_common/form.html.twig' %}
+
+{% block body %}
+    {% if user %}
+        <style>
+            label.required:after {
+                content: ' *'
+            }
+        </style>
+        <h1 class="title">Edit event {{ event.name }}</h1>
```
Comment (akrabat) [inline app/templates/Event/edit.html.twig]: Don't need "event" here: just "Edit {event's name}" is fine

File: app/src/Event/EventController.php [hunk 4]
```diff
@@ -291,6 +302,85 @@ public function submit()
...
+     *
+     * @return bool
+     */
+    protected function userIsAllowedToEdit($event)
+    {
+        if (! isset($_SESSION['user'])) {
+            return false;
+        }
+
+        if (isset($event['can_edit'])) {
```
Comment (akrabat) [inline app/src/Event/EventController.php]: $event is an instance of EventEntity here, so we can't access it like an array. You'll need a `getCanEdit()` method on the EventEntity.

File: app/src/Event/EventController.php [hunk 5]
```diff
@@ -291,6 +302,85 @@ public function submit()
...
+            $this->redirectToListPage();
+        }
+
+        if (! $this->userIsAllowedToEdit($event)) {
+            $this->redirectToDetailPage($event->getUrlFriendlyName());
+        }
+
+        /** @var FormFactoryInterface $factory */
+        $factory = $this->application->formFactory;
+        $form    = $factory->create(new EventFormType(), $event->setTags(implode(', ', $event->getTags())));
```
Comment (akrabat) [inline app/src/Event/EventController.php]: I'm not happy about rewriting the way tags are stored internally from an array to a csv.
I think that the correct Symfony2 Forms solution to this problem is to implement a "DataTransformer" like we did for dates here: https://github.com/joindin/joindin-web2/blob/master/app/src/Form/DataTransformer/DateTransformer.php

**Comment by akrabat** [comment]
On testing, I filled in every field and pressed submit and got this error:

    Type: Symfony\Component\PropertyAccess\Exception\NoSuchPropertyException
    Message: Neither the property "location" nor one of the methods "addLocation()"/"removeLocation()", "setLocation()", "location()", "__set()" or "__call()" exist and have public access in class "Event\EventEntity".
    File: /vagrant/joindin-web2/vendor/symfony/property-access/Symfony/Component/PropertyAccess/PropertyAccessor.php
    Line: 456

**Comment by akrabat** [comment]
When editing an event where I had set the URL to "http://example.com" via web1, the edit form on web2 displays "http://api.dev.joind.in/v2.1/events/53" in the "Website URL" form field.

**Comment by heiglandreas** [comment]
This is superseded by PR #210

**Pull request closed without merging**